### PR TITLE
Nameopt fixes

### PR DESCRIPTION
**Pull request by beldmit**
This pull request unifies -nameopt flag processing in command-line apps and reduces usage of the X509_NAME_online function.

### CODE REVIEW

File: apps/apps.c [hunk 1]
```diff
@@ -151,18 +151,10 @@ int ctx_set_ctlog_list_file(SSL_CTX *ctx, const char *path)
...
-    BIO_puts(out, p);
-    OPENSSL_free(p);
-
-    p = X509_NAME_oneline(X509_get_issuer_name(x), NULL, 0);
-    BIO_puts(out, "\nissuer=");
-    BIO_puts(out, p);
-    BIO_puts(out, "\n");
-    OPENSSL_free(p);
+    print_name(out, "subject=", X509_get_subject_name(x), get_nameopt());
+		BIO_puts(out, "\n");
```
Comment (richsalz) [inline apps/apps.c]: extra indent here and on the next puts line.

File: apps/apps.c [hunk 2]
```diff
@@ -966,6 +958,24 @@ int load_crls(const char *file, STACK_OF(X509_CRL) **crls, int format,
 #define X509_FLAG_CA (X509_FLAG_NO_ISSUER | X509_FLAG_NO_PUBKEY | \
                          X509_FLAG_NO_HEADER | X509_FLAG_NO_VERSION)
 
+static unsigned long nmflag = 0;
+static char nmflag_set = 0;
+
+int set_nameopt(const char *arg)
+{
+  int ret = set_name_ex(&nmflag, arg);
+  if (ret)
```
Comment (richsalz) [inline apps/apps.c]: blank line after the variable init, and remove the curly braces.

File: apps/s_client.c [hunk 3]
```diff
@@ -2859,10 +2858,12 @@ static void print_stuff(BIO *bio, SSL *s, int full)
             /* Redundant if we showed the whole chain */
             if (!(c_showcerts && got_a_chain))
                 PEM_write_bio_X509(bio, peer);
-            X509_NAME_oneline(X509_get_subject_name(peer), buf, sizeof buf);
-            BIO_printf(bio, "subject=%s\n", buf);
-            X509_NAME_oneline(X509_get_issuer_name(peer), buf, sizeof buf);
-            BIO_printf(bio, "issuer=%s\n", buf);
+            BIO_puts(bio, "subject=");
```
Comment (richsalz) [inline apps/s_client.c]: should we have a utility routine in apps that wraps the prefix/name/suffix?  Something like:
```
app_print(const char* prefix, X509_NAME* n, const char* suffix)
{
print prefix
call name_oneline
print suffix
}
```

File: apps/apps.c (hunk 1 above)
Comment (beldmit) [inline apps/apps.c]: Fixed.

File: apps/apps.c (hunk 2 above)
Comment (beldmit) [inline apps/apps.c]: Fixed.

File: apps/s_client.c (hunk 3 above)
Comment (beldmit) [inline apps/s_client.c]: It makes sense though I am not sure this function will fit all the necessary places.

Comment (richsalz) [inline apps/s_client.c]: okay.  I'll think about it.  Not for this PR.

**Comment by richsalz** [comment]
@beldmit look at the travis failures; you need a declaration in apps.h or such.

**Comment by beldmit** [comment]
Fixed now.

**Comment by richsalz** [comment]
Great, just need another reviewer and I'll merge.

**Comment by beldmit** [comment]
I provided support and test for Unicode output in crl -text command. Could you please review it again?

**Comment by beldmit** [comment]
In fact the -nameopt key was documented for the crl utility. It just did not work in case of -text option.

**Comment by richsalz** [comment]
I meant the new API for CRL print

**Comment by beldmit** [comment]
I'm afraid not because the old API seems to be undocumented too...

**Comment by richsalz** [comment]
sigh.  okay.  still approved by me...

File: apps/apps.c [hunk 4]
```diff
@@ -149,20 +149,30 @@ int ctx_set_ctlog_list_file(SSL_CTX *ctx, const char *path)
...
+
+int set_nameopt(const char *arg)
 {
-    char *p;
+  int ret = set_name_ex(&nmflag, arg);
+
+  if (ret) 
+    nmflag_set = 1;
+		
+	return ret;
```
Comment (levitte) [inline apps/apps.c]: Sorry, code style wise, this indentation isn't acceptable.  We use 4 space indent, but what got me was the indentation if the return statement.

File: apps/apps.c [hunk 5]
```diff
@@ -149,20 +149,30 @@ int ctx_set_ctlog_list_file(SSL_CTX *ctx, const char *path)
...
+}
 
-    p = X509_NAME_oneline(X509_get_subject_name(x), NULL, 0);
-    BIO_puts(out, "subject=");
-    BIO_puts(out, p);
-    OPENSSL_free(p);
+unsigned long get_nameopt(void)
+{
+  return (nmflag_set) ? nmflag : XN_FLAG_ONELINE;
```
Comment (levitte) [inline apps/apps.c]: ... but since I already brought it up, 4 space indent.

File: apps/s_client.c [hunk 6]
```diff
@@ -2859,10 +2858,12 @@ static void print_stuff(BIO *bio, SSL *s, int full)
...
+            X509_NAME_print_ex(bio, X509_get_subject_name(peer), 0, get_nameopt());
+            BIO_puts(bio, "\n");
+            BIO_puts(bio, "issuer=");
+            X509_NAME_print_ex(bio, X509_get_issuer_name(peer), 0, get_nameopt());
+            BIO_puts(bio, "\n");
```
Comment (levitte) [inline apps/s_client.c]: Why don't you use `print_name` here?

File: apps/s_server.c [hunk 7]
```diff
@@ -2688,10 +2688,12 @@ static void print_connection_info(SSL *con)
...
-        X509_NAME_oneline(X509_get_subject_name(peer), buf, sizeof buf);
-        BIO_printf(bio_s_out, "subject=%s\n", buf);
-        X509_NAME_oneline(X509_get_issuer_name(peer), buf, sizeof buf);
-        BIO_printf(bio_s_out, "issuer=%s\n", buf);
+        BIO_puts(bio_s_out, "subject=");
+        X509_NAME_print_ex(bio_s_out, X509_get_subject_name(peer), 0, get_nameopt());
+        BIO_puts(bio_s_out, "\n");
+        BIO_puts(bio_s_out, "issuer=");
+        X509_NAME_print_ex(bio_s_out, X509_get_issuer_name(peer), 0, get_nameopt());
+        BIO_puts(bio_s_out, "\n");
```
Comment (levitte) [inline apps/s_server.c]: `print_name`?

File: apps/x509.c [hunk 8]
```diff
@@ -726,13 +721,13 @@ int x509_main(int argc, char **argv)
...
-
-                X509_NAME_oneline(X509_get_issuer_name(x), buf, sizeof buf);
-                BIO_printf(out, " * Issuer:  %s\n"
-                                " */\n", buf);
+                                " * Subject: ");
+                X509_NAME_print_ex(out, X509_get_subject_name(x), 0, get_nameopt());
+                BIO_puts(out, "\n");
+                BIO_printf(out, " * Issuer:  ");
+                X509_NAME_print_ex(out, X509_get_issuer_name(x), 0, get_nameopt());
+                BIO_puts(out, "\n */\n");
```
Comment (levitte) [inline apps/x509.c]: I imagine you could use `print_name` here as well...

File: apps/apps.c (hunk 4 above)
Comment (beldmit) [inline apps/apps.c]: Fixed

**Comment by richsalz** [comment]
@beldmit, thanks for your work to help make openssl more language-neutral! commit b5c4209.

**Pull request closed without merging**